File: fornituras-api/src/main/java/com/numobiz/solutions/fornituras/modules/warehouses/service/WarehouseService.java
package com.numobiz.solutions.fornituras.modules.warehouses.service;

import com.numobiz.solutions.fornituras.common.audit.AuditWriter;
import com.numobiz.solutions.fornituras.common.exception.BadRequestException;
import com.numobiz.solutions.fornituras.common.exception.ConflictException;
import com.numobiz.solutions.fornituras.common.exception.NotFoundException;
import com.numobiz.solutions.fornituras.common.text.NameNormalizer;
import com.numobiz.solutions.fornituras.modules.catalog.CatalogCodes;
import com.numobiz.solutions.fornituras.modules.catalog.service.CatalogService;
import com.numobiz.solutions.fornituras.modules.users.repository.UserRepository;
import com.numobiz.solutions.fornituras.modules.warehouses.dto.WarehouseCreateRequest;
import com.numobiz.solutions.fornituras.modules.warehouses.dto.WarehouseDetail;
import com.numobiz.solutions.fornituras.modules.warehouses.dto.WarehouseSummary;
import com.numobiz.solutions.fornituras.modules.warehouses.entity.Warehouse;
import com.numobiz.solutions.fornituras.modules.warehouses.mapper.WarehouseMapper;
import com.numobiz.solutions.fornituras.modules.warehouses.repository.WarehouseRepository;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Transactional(readOnly = true)
public class WarehouseService {

	private final WarehouseRepository repository;
	private final WarehouseMapper mapper;
	private final WarehouseUsageQuery usageQuery;
	private final CatalogService catalogService;
	private final UserRepository userRepository;
	private final AuditWriter audit;

	public WarehouseService(
			WarehouseRepository repository,
			WarehouseMapper mapper,
			WarehouseUsageQuery usageQuery,
			CatalogService catalogService,
			UserRepository userRepository,
			AuditWriter audit) {
		this.repository = repository;
		this.mapper = mapper;
		this.usageQuery = usageQuery;
		this.catalogService = catalogService;
		this.userRepository = userRepository;
		this.audit = audit;
	}

	public Page<WarehouseSummary> findAll(Boolean active, Long tipoItemId, Pageable pageable) {
		Page<Warehouse> page;
		if (active != null && tipoItemId != null) {
			page = repository.findByActiveAndTipoItemId(active, tipoItemId, pageable);
		} else if (active != null) {
			page = repository.findByActive(active, pageable);
		} else if (tipoItemId != null) {
			page = repository.findByTipoItemId(tipoItemId, pageable);
		} else {
			page = repository.findAll(pageable);
		}
		return page.map(w -> mapper.toSummary(w, catalogService.resolveName(w.getTipoItemId())));
	}

	public WarehouseDetail findById(Long id) {
		Warehouse warehouse = getOrThrow(id);
		return toDetail(warehouse);
	}

	@Transactional
	public WarehouseDetail create(WarehouseCreateRequest request) {
		String normalized = NameNormalizer.normalize(request.nombre());
		String codigo = request.codigo().trim();

		if (repository.existsByCodigoIgnoreCase(codigo)) {
			throw new ConflictException("Ya existe un almacén con la clave: " + codigo);
		}
		if (repository.existsByNombreNormalizado(normalized)) {
			throw new ConflictException("Ya existe un almacén con el nombre: " + request.nombre());
		}
		validateReferences(request);

		Warehouse warehouse = new Warehouse();
		warehouse.setCodigo(codigo);
		warehouse.setNombreNormalizado(normalized);
		apply(warehouse, request);
		warehouse.setActive(true);

		Warehouse saved = repository.save(warehouse);
		audit.record("CREATE_WAREHOUSE", saved.getId());
		return toDetail(saved);
	}

	@Transactional
	public WarehouseDetail update(Long id, WarehouseCreateRequest request) {
		Warehouse warehouse = getOrThrow(id);
		String normalized = NameNormalizer.normalize(request.nombre());
		String codigo = request.codigo().trim();

		repository.findByCodigoIgnoreCase(codigo)
				.filter(existing -> !existing.getId().equals(id))
				.ifPresent(existing -> {
					throw new ConflictException("Ya existe un almacén con la clave: " + codigo);
				});
		repository.findByNombreNormalizado(normalized)
				.filter(existing -> !existing.getId().equals(id))
				.ifPresent(existing -> {
					throw new ConflictException("Ya existe un almacén con el nombre: " + request.nombre());
				});
		validateReferences(request);

		warehouse.setCodigo(codigo);
		warehouse.setNombreNormalizado(normalized);
		apply(warehouse, request);

		Warehouse saved = repository.save(warehouse);
		audit.record("UPDATE_WAREHOUSE", saved.getId());
		return toDetail(saved);
	}

	/**
	 * Desactivación lógica (nunca borrado físico): un almacén en uso seguiría siendo referenciado,
	 * así que solo se marca inactivo y deja de ofrecerse como ubicación (001) o destino (007).
	 */
	@Transactional
	public void deactivate(Long id) {
		Warehouse warehouse = getOrThrow(id);
		warehouse.setActive(false);
		repository.save(warehouse);
		audit.record("DEACTIVATE_WAREHOUSE", id);
	}

	/**
	 * Borrado físico solo si el almacén NO está en uso. Si tiene fornituras o traslados asociados,
	 * se bloquea y se ofrece la desactivación en su lugar (integridad referencial, FR-003).
	 */
	@Transactional
	public void delete(Long id) {
		Warehouse warehouse = getOrThrow(id);
		if (usageQuery.countUsage(id) > 0) {
			throw new ConflictException(
					"El almacén tiene fornituras o traslados asociados; desactívelo en vez de eliminarlo.");
		}
		repository.delete(warehouse);
		audit.record("DELETE_WAREHOUSE", id);
	}

	private void apply(Warehouse warehouse, WarehouseCreateRequest request) {
		warehouse.setNombre(request.nombre().trim());
		warehouse.setTipoItemId(request.tipoItemId());
		warehouse.setMunicipio(blankToNull(request.municipio()));
		warehouse.setEstado(blankToNull(request.estado()));
		warehouse.setDireccion(request.direccion());
		warehouse.setCp(request.cp());
		warehouse.setLatitud(request.latitud());
		warehouse.setLongitud(request.longitud());
		warehouse.setResponsableId(request.responsableId());
		warehouse.setTelefono(request.telefono());
		warehouse.setEmailContacto(request.emailContacto());
		warehouse.setCapacidad(request.capacidad());
		warehouse.setObservaciones(request.observaciones());
	}

	private void validateReferences(WarehouseCreateRequest request) {
		catalogService.requireActiveItem(request.tipoItemId(), CatalogCodes.TIPO_ALMACEN);
		if (request.responsableId() != null && !userRepository.existsById(request.responsableId())) {
			throw new BadRequestException("Usuario responsable no encontrado: " + request.responsableId());
		}
	}

	private String blankToNull(String value) {
		return (value == null || value.isBlank()) ? null : value.trim();
	}

	private WarehouseDetail toDetail(Warehouse warehouse) {
		long ocupacion = usageQuery.countUsage(warehouse.getId());
		Double porcentaje = (warehouse.getCapacidad() != null && warehouse.getCapacidad() > 0)
				? (ocupacion * 100.0 / warehouse.getCapacidad())
				: null;
		String tipoNombre = catalogService.resolveName(warehouse.getTipoItemId());
		return mapper.toDetail(warehouse, tipoNombre, ocupacion, porcentaje);
	}

	private Warehouse getOrThrow(Long id) {
		return repository.findById(id)
				.orElseThrow(() -> new NotFoundException("Almacén no encontrado: " + id));
	}
}

File: fornituras-api-dotnet/src/Fornituras.Api/Services/WarehouseService.cs
using Fornituras.Api.Common;
using Fornituras.Api.Common.Text;
using Fornituras.Api.Data;
using Fornituras.Api.Data.Entities;
using Fornituras.Api.Dto;
using Microsoft.EntityFrameworkCore;
using EntityEquipmentStatus = Fornituras.Api.Data.Entities.EquipmentStatus;

namespace Fornituras.Api.Services;

public sealed class WarehouseService(
    ApplicationDbContext db,
    CatalogService catalogService,
    IAuditWriter audit) : IWarehouseService
{
    public Task<PageResult<WarehouseSummary>> FindAllAsync(
        bool? active,
        long? tipoItemId,
        PaginationQuery pagination,
        CancellationToken cancellationToken = default) =>
        FindAllInternalAsync(active, tipoItemId, pagination.Page, pagination.Size, cancellationToken);

    private async Task<PageResult<WarehouseSummary>> FindAllInternalAsync(
        bool? active,
        long? tipoItemId,
        int page,
        int size,
        CancellationToken cancellationToken = default)
    {
        var query = db.Warehouses.AsNoTracking();
        if (active.HasValue)
        {
            query = query.Where(w => w.Active == active.Value);
        }

        if (tipoItemId.HasValue)
        {
            query = query.Where(w => w.TipoItemId == tipoItemId.Value);
        }

        query = query.OrderBy(w => w.Nombre);
        var total = await query.CountAsync(cancellationToken);
        var warehouses = await query.Skip(page * size).Take(size).ToListAsync(cancellationToken);
        var tipoNames = await catalogService.ResolveNamesAsync(
            warehouses.Select(w => w.TipoItemId),
            cancellationToken);

        var summaries = warehouses.Select(w => new WarehouseSummary(
            w.Id,
            w.Codigo,
            w.Nombre,
            w.TipoItemId,
            tipoNames.GetValueOrDefault(w.TipoItemId, "?"),
            w.Active)).ToList();

        return PageResult<WarehouseSummary>.From(summaries, total, page, size);
    }

    public async Task<WarehouseDetail> FindByIdAsync(long id, CancellationToken cancellationToken = default)
    {
        var warehouse = await db.Warehouses.AsNoTracking()
            .FirstOrDefaultAsync(w => w.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Warehouse not found: {id}");

        var tipoNombre = await catalogService.ResolveNameAsync(warehouse.TipoItemId, cancellationToken) ?? "?";
        var ocupacion = await CountOccupancyAsync(id, cancellationToken);
        double? porcentaje = warehouse.Capacidad is > 0
            ? ocupacion * 100.0 / warehouse.Capacidad.Value
            : null;

        return new WarehouseDetail(
            warehouse.Id,
            warehouse.Codigo,
            warehouse.Nombre,
            warehouse.TipoItemId,
            tipoNombre,
            warehouse.Municipio,
            warehouse.Estado,
            warehouse.Direccion,
            warehouse.Cp,
            warehouse.Latitud,
            warehouse.Longitud,
            warehouse.ResponsableId,
            warehouse.Telefono,
            warehouse.EmailContacto,
            warehouse.Capacidad,
            warehouse.Observaciones,
            warehouse.Active,
            ocupacion,
            porcentaje,
            warehouse.CreatedAt,
            warehouse.UpdatedAt);
    }

    public async Task<WarehouseDetail> CreateAsync(
        WarehouseCreateRequest request,
        CancellationToken cancellationToken = default)
    {
        await catalogService.RequireActiveItemAsync(request.TipoItemId, CatalogCodes.TipoAlmacen, cancellationToken);
        await EnsureUniqueCodigoAsync(request.Codigo, null, cancellationToken);
        await EnsureUniqueNombreAsync(request.Nombre, null, cancellationToken);

        if (request.ResponsableId.HasValue &&
            !await db.Users.AnyAsync(u => u.Id == request.ResponsableId.Value, cancellationToken))
        {
            throw new BadRequestException($"Usuario responsable no encontrado: {request.ResponsableId}");
        }

        var now = DateTime.UtcNow;
        var warehouse = MapFromRequest(request);
        warehouse.Active = true;
        warehouse.CreatedAt = now;
        warehouse.UpdatedAt = now;

        db.Warehouses.Add(warehouse);
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("CREATE_WAREHOUSE", warehouse.Id, cancellationToken);
        return await FindByIdAsync(warehouse.Id, cancellationToken);
    }

    public async Task<WarehouseDetail> UpdateAsync(
        long id,
        WarehouseCreateRequest request,
        CancellationToken cancellationToken = default)
    {
        var warehouse = await db.Warehouses.FirstOrDefaultAsync(w => w.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Warehouse not found: {id}");

        await catalogService.RequireActiveItemAsync(request.TipoItemId, CatalogCodes.TipoAlmacen, cancellationToken);
        await EnsureUniqueCodigoAsync(request.Codigo, id, cancellationToken);
        await EnsureUniqueNombreAsync(request.Nombre, id, cancellationToken);

        if (request.ResponsableId.HasValue &&
            !await db.Users.AnyAsync(u => u.Id == request.ResponsableId.Value, cancellationToken))
        {
            throw new BadRequestException($"Usuario responsable no encontrado: {request.ResponsableId}");
        }

        ApplyRequest(warehouse, request);
        warehouse.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("UPDATE_WAREHOUSE", warehouse.Id, cancellationToken);
        return await FindByIdAsync(warehouse.Id, cancellationToken);
    }

    public async Task DeactivateAsync(long id, CancellationToken cancellationToken = default)
    {
        var warehouse = await db.Warehouses.FirstOrDefaultAsync(w => w.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Warehouse not found: {id}");

        warehouse.Active = false;
        warehouse.UpdatedAt = DateTime.UtcNow;
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("DEACTIVATE_WAREHOUSE", warehouse.Id, cancellationToken);
    }

    public async Task DeleteAsync(long id, CancellationToken cancellationToken = default)
    {
        var usage = await CountUsageAsync(id, cancellationToken);
        if (usage > 0)
        {
            throw new ConflictException(
                "No se puede eliminar el almacén porque tiene fornituras o traslados asociados. Desactívelo.");
        }

        var warehouse = await db.Warehouses.FirstOrDefaultAsync(w => w.Id == id, cancellationToken)
            ?? throw new NotFoundException($"Warehouse not found: {id}");

        db.Warehouses.Remove(warehouse);
        await db.SaveChangesAsync(cancellationToken);
        await audit.RecordAsync("DELETE_WAREHOUSE", id, cancellationToken);
    }

    private async Task EnsureUniqueCodigoAsync(string codigo, long? excludeId, CancellationToken cancellationToken)
    {
        var normalized = codigo.Trim().ToUpperInvariant();
        var query = db.Warehouses.Where(w => w.Codigo.ToUpper() == normalized);
        if (excludeId.HasValue)
        {
            query = query.Where(w => w.Id != excludeId.Value);
        }

        if (await query.AnyAsync(cancellationToken))
        {
            throw new ConflictException($"Ya existe un almacén con el código '{codigo.Trim()}'.");
        }
    }

    private async Task EnsureUniqueNombreAsync(string nombre, long? excludeId, CancellationToken cancellationToken)
    {
        var normalized = NameNormalizer.Normalize(nombre);
        var query = db.Warehouses.Where(w => w.NombreNormalizado == normalized);
        if (excludeId.HasValue)
        {
            query = query.Where(w => w.Id != excludeId.Value);
        }

        if (await query.AnyAsync(cancellationToken))
        {
            throw new ConflictException($"Ya existe un almacén con el nombre '{nombre.Trim()}'.");
        }
    }

    private async Task<long> CountOccupancyAsync(long warehouseId, CancellationToken cancellationToken) =>
        await db.Equipment.CountAsync(
            e => e.WarehouseId == warehouseId && e.Status != EntityEquipmentStatus.BAJA_DEFINITIVA,
            cancellationToken);

    private async Task<long> CountUsageAsync(long warehouseId, CancellationToken cancellationToken)
    {
        var equipmentCount = await db.Equipment.CountAsync(e => e.WarehouseId == warehouseId, cancellationToken);
        var transferCount = await db.Transfers.CountAsync(
            t => t.OrigenId == warehouseId || t.DestinoId == warehouseId,
            cancellationToken);
        return equipmentCount + transferCount;
    }

    private static Warehouse MapFromRequest(WarehouseCreateRequest request)
    {
        var warehouse = new Warehouse();
        ApplyRequest(warehouse, request);
        return warehouse;
    }

    private static void ApplyRequest(Warehouse warehouse, WarehouseCreateRequest request)
    {
        warehouse.Codigo = request.Codigo.Trim();
        warehouse.Nombre = request.Nombre.Trim();
        warehouse.NombreNormalizado = NameNormalizer.Normalize(request.Nombre);
        warehouse.TipoItemId = request.TipoItemId;
        warehouse.Municipio = request.Municipio?.Trim();
        warehouse.Estado = request.Estado?.Trim();
        warehouse.Direccion = request.Direccion?.Trim();
        warehouse.Cp = request.Cp?.Trim();
        warehouse.Latitud = request.Latitud;
        warehouse.Longitud = request.Longitud;
        warehouse.ResponsableId = request.ResponsableId;
        warehouse.Telefono = request.Telefono?.Trim();
        warehouse.EmailContacto = request.EmailContacto?.Trim();
        warehouse.Capacidad = request.Capacidad;
        warehouse.Observaciones = request.Observaciones?.Trim();
    }
}
